### PR TITLE
Handling timers for repeat dispute participation requests

### PR DESCRIPTION
**Pull request by BradleyOlson64**
This is a follow up to https://github.com/paritytech/polkadot/pull/6838 aiming to make the metric `polkadot_parachain_dispute_participation_pipeline_durations` more accurate. 

Previously, if multiple participation requests were made for the same candidate then the older request would simply be dropped. But dropping the older request meant dropping its request timer as well. The pipeline duration metric only works properly if each request timer is dropped after we've participated in its corresponding dispute. 

These changes fix the issue, keeping the oldest timer for each participation request alive and disposing of newer duplicate timers using `stop_and_discard()`

### CODE REVIEW

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 1]
```diff
@@ -151,7 +146,7 @@ impl PartialEq for ParticipationRequest {
 			candidate_receipt,
 			candidate_hash,
 			session: _session,
```
Comment (mrcnski) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:148]: nit: Unused `_session`.

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 2]
```diff
@@ -231,15 +226,31 @@ impl Queues {
...
 		if priority.is_priority() {
 			if self.priority.len() >= PRIORITY_QUEUE_SIZE {
 				return Err(QueueError::PriorityFull)
 			}
-			// Remove any best effort entry:
-			self.best_effort.remove(&comparator);
-			self.priority.insert(comparator, req);
+			// Remove any best effort entry, using it to replace our new
+			// request. This preserves the original request timer from
+			// the older request.
```
Comment (mrcnski) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:237]: nit: I find this function hard to follow now, with all the nested ifs and elses. It would help to have a docstring for the function. Also, while the comments are awesome (⭐️), they are a bit repetitive, and maybe can be moved to the docstring. i.e. I envision something like:

```
/// # Request timers
///
/// [`ParticipationRequest`]s contain request timers. 
/// Where an old request would be replaced by a new one, we try to keep the old request.
/// This prevents request timers from resetting on each new request.
```

so the comment here becomes

```suggestion
			// Remove any best effort entry, using it to replace our new
			// request.
```

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 3]
```diff
@@ -231,15 +226,31 @@ impl Queues {
...
+					timer.stop_and_discard();
+				}
+				req = older_request;
+			}
+			// Keeping old request if any. This prevents request timers
+			// from resetting on each new request at the same priority
+			// level for the same candidate.
+			if self.priority.contains_key(&comparator) {
+				if let Some(timer) = req.request_timer {
+					timer.stop_and_discard();
```
Comment (mrcnski) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:249]: Doesn't this stop an old request from the best effort queue? Not sure I'm totally following the logic though -- I think a docstring for the function would help me. 🙂

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 4]
```diff
@@ -251,7 +262,16 @@ impl Queues {
...
-			self.best_effort.insert(comparator, req);
+			// Keeping old request if any. This prevents request timers
+			// from resetting on each new request at the same priority
+			// level for the same candidate.
+			if self.best_effort.contains_key(&comparator) {
+				if let Some(timer) = req.request_timer {
+					timer.stop_and_discard();
+				}
+			} else {
+				self.best_effort.insert(comparator, req);
```
Comment (eskimor) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:273]: You can avoid the double lookup via [`entry`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry).

File: node/core/dispute-coordinator/src/participation/queues/mod.rs (hunk 3 above)
Comment (BradleyOlson64) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:249]: I think you're reading the code right. If there were identical old requests in both the best effort and priority queues then the old best effort request would be thrown out.

But in practice a request should never be in both the best effort queue and the priority queue at the same time. We remove best effort queue entries to ensure this.

If by some fluke a request ends up in both queues it wouldn't be the end of the world. We still get decent data, since we still have the property that no timers send metric data before we've participated in the dispute.

File: node/core/dispute-coordinator/src/participation/queues/mod.rs (hunk 4 above)
Comment (BradleyOlson64) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:273]: Maybe I'm missing something obvious, but here's the problem I see there. The intuitive pattern is something like:

`self.priority.entry(comparator).or_insert(req);`

But this pattern drops `req` without calling `stop_and_discard()` on its timer if there was an older request. This sends bad data to our metrics. I don't see a way to call `stop_and_discard()` without using a double lookup. 🤷‍♂️

**Review by mrcnski — APPROVED** [review]
Great docstring!

File: node/core/dispute-coordinator/src/participation/queues/mod.rs (hunk 4 above)
Comment (alexgparity) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:276]: Can't you use [or_insert_with](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_insert_with) to execute `timer.stop_and_discard()` first and then insert `req`?

Comment (eskimor) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:276]: you can also always pattern match on the `Entry` constructors and do whatever you need to do. See [here](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/statement-table/src/generic.rs#L391) for an example.

Comment (BradleyOlson64) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs]: Your solution worked @eskimor Much appreciated!

**Review by eskimor — APPROVED** [review]
